README.md
# DPFlickView

A simple and easy to use subclass of UIView that lets you create a Tinder-like flick view.

## Usage

Import the header file where you want to use DPFlickView

```objc
#import <DPFLickView.h>
```

After initializing a new DPFlickView

```objc
DPFlickView *flickView = [[DPFlickView alloc] initWithFrame:CGRectMake(70, 150, 180, 180)];
flickView.backgroundColor = [UIColor blueColor];
[self.view addSubview flickView];
```

You can assign one action when the view is flicked to the right and to the left by passing a block.

```objc
flickView.leftAction = ^{ NSLog(@"LEFT ACTION");};
flickView.rightAction = ^{ NSLog(@"RIGHT ACTION");};
```

It's also possible to pass an `NSArray` to indicate which subviews should fade when flicking left and right.

```objc
flickView.leftFadingSubviews = @[mySubview, myOtherSubview];
flickView.rightFadingSubviews = @[statusLabel, avatarImage];
```

You can set a custom rotation angle in degrees by setting the `rotationAngle` property to an int, the default value is 10 degrees:

```objc
flickView.rotationAngle = 25;
```

The `alphaValue` property is `readonly` and allows you to combine the fading that's happening in the flick view together with events outside it.


## Installation

DPFlickView is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "DPFlickView"

## Author

David Harver Pollak, bebus77@gmail.com

[@bebus77](https://twitter.com/bebus77)

## License

DPFlickView is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# DPFlickView CHANGELOG

## 0.1.0

Initial release.


Classes/ios/DPFlickView.h
//
//  DPFlickView.h
//  flickview
//
//  Created by David Harver Pollak on 16/01/14.
//  Copyright (c) 2014 David Harver Pollak. All rights reserved.
//

#import <UIKit/UIKit.h>

@interface DPFlickView : UIView {
    CGPoint initialTouchPoint;
    CGPoint initialCenterPosition;
    CGPoint initialScreenPoint;
    int rotationDirection;
}

// Angle of rotation during the flick
@property (nonatomic) int rotationAngle;

// Incremental opacity value applied to selected subviews
@property (readonly) CGFloat alphaValue;

// Subviews to be faded while flicking left
@property (nonatomic, strong) NSArray *leftFadingSubviews;

// Subviews to be faded while flicking right
@property (nonatomic, strong) NSArray *rightFadingSubviews;

// Block to be triggered when the view is triggered left
@property (nonatomic, copy) void (^leftAction)(void);

// Block to be triggered when the view is triggered right
@property (nonatomic, copy) void (^rightAction)(void);

- (id)initWithFrame:(CGRect)frame;

@end


Classes/ios/DPFlickView.m
//
//  DPFlickView.m
//  flickview
//
//  Created by David Harver Pollak on 16/01/14.
//  Copyright (c) 2014 David Harver Pollak. All rights reserved.
//

#import "DPFlickView.h"

@implementation DPFlickView

#define DEGREES_TO_RADIANS(x) (M_PI * (x) / 180.0)
const int kRotationRadius = 180.0;

- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        // Default rotation angle defined at 10 degrees
        self.rotationAngle = 10;
        
        // Setup initial center position to restore it afterwards
        initialCenterPosition = self.center;
        
        // Initial alpha value for the subviews' opacity
        _alphaValue = 1.0;
        
        // Default initialization of the left fading subviews
        self.leftFadingSubviews = @[];
        
        // Default initialization of the right fading subviews
        self.rightFadingSubviews = @[];
        
        // Default block triggered when the view is flicked to left
        self.leftAction = ^{};
        
        // Default block triggered when the view is flicked to right
        self.rightAction = ^{};
    }
    return self;
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    // Get the initial touch relative to the view itself
    initialTouchPoint = [[touches anyObject] locationInView:self];
    
    // Get the initial touch relative to the view's superview
    initialScreenPoint = [[touches anyObject] locationInView:self.superview];
    
    // Set the rotation direction based on the y value of the initial touch
    rotationDirection = -1;
    if (initialTouchPoint.y < self.bounds.size.height / 2) {
        rotationDirection = 1;
    }
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    // Get the latest touch
    CGPoint movedTouchPoint = [[touches anyObject] locationInView:self.superview];
    
    // Set the new view's center based on the moved touch
    self.center = CGPointMake(movedTouchPoint.x - initialTouchPoint.x + self.bounds.size.width / 2,
                              movedTouchPoint.y - initialTouchPoint.y + self.bounds.size.height / 2);
    
    // Incremental angle setup
    double angle = (self.center.x - kRotationRadius) / kRotationRadius * DEGREES_TO_RADIANS(self.rotationAngle);
    
    // Incremental opacity alpha value
    _alphaValue = fabsf((self.center.x - kRotationRadius) / kRotationRadius);
    
    // Check the view's direction
    if (movedTouchPoint.x <= initialScreenPoint.x) {
        [self fadeSubviews:self.leftFadingSubviews];
    } else {
        [self fadeSubviews:self.rightFadingSubviews];
    }
    
    // Apply the rotation transformation to the view
    [self setTransform:CGAffineTransformMakeRotation(rotationDirection * angle)];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    // Restore initial position
    CGPoint destinationCenter = initialCenterPosition;
    
    // Check if one of the to blocks should be triggered
    if (self.center.x < self.superview.center.x / 2) {
        destinationCenter = CGPointMake(self.superview.bounds.origin.x - self.bounds.size.width, self.center.y);
        self.leftAction();
    } else if (self.center.x > self.superview.center.x * 3 / 2) {
        destinationCenter = CGPointMake(self.superview.bounds.size.width + self.bounds.size.width, self.center.y);
        self.rightAction();
    }
    
    // Animate, move and restore the opacity alpha value
    [UIView animateWithDuration:0.2 animations:^{
        self.center = destinationCenter;
        [self setTransform:CGAffineTransformMakeRotation(DEGREES_TO_RADIANS(0))];
        
        _alphaValue = 0.0;
        [self fadeSubviews:self.leftFadingSubviews];
        [self fadeSubviews:self.rightFadingSubviews];
    }];
}

- (void)fadeSubviews:(NSArray *)sViews {
    // Set the opacity alpha value for each view in the given array
    for (UIView *sView in sViews) {
        // Check if the current element in the array is a UIView and subview of self
        if ([sView isKindOfClass:[UIView class]] && [sView isDescendantOfView:self]) {
            sView.alpha = self.alphaValue;
        }
    }
}

@end


Project/FlickViewDemo/FlickViewDemo/AppDelegate.m
//
//  AppDelegate.m
//  FlickViewDemo
//
//  Created by David Harver Pollak on 25/01/14.
//  Copyright (c) 2014 David Harver Pollak. All rights reserved.
//

#import "AppDelegate.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Project/FlickViewDemo/FlickViewDemo/ViewController.m
//
//  ViewController.m
//  FlickViewDemo
//
//  Created by David Harver Pollak on 25/01/14.
//  Copyright (c) 2014 David Harver Pollak. All rights reserved.
//

#import "ViewController.h"
#import <DPFlickView.h>

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    DPFlickView *flickView = [[DPFlickView alloc] initWithFrame:CGRectMake(70, 150, 180, 180)];
    flickView.backgroundColor = [UIColor blueColor];
    [self.view addSubview:flickView];
    
    flickView.leftAction = ^{ NSLog(@"LEFT ACTION");};
    flickView.rightAction = ^{ NSLog(@"RIGHT ACTION");};
    
    UILabel *rightLabel = [[UILabel alloc] initWithFrame:CGRectMake(10, 10, 60, 15)];
    rightLabel.textColor = [UIColor whiteColor];
    rightLabel.text = @"RIGHT";
    rightLabel.alpha = 0.0;
    [flickView addSubview:rightLabel];
    
    UILabel *leftLabel = [[UILabel alloc] initWithFrame:CGRectMake(130, 10, 50, 15)];
    leftLabel.textColor = [UIColor whiteColor];
    leftLabel.text = @"LEFT";
    leftLabel.alpha = 0.0;
    [flickView addSubview:leftLabel];
    
    flickView.leftFadingSubviews = @[leftLabel];
    flickView.rightFadingSubviews = @[rightLabel];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end
